include/swift/SILOptimizer/Analysis/VisitBarrierAccessScopes.h

```diff
@@ -0,0 +1,190 @@
+class SILInstruction;
+class SILBasicBlock;
+class SILInstruction;
+
+/// Visits the begin_access instructions that must be regarded as barriers
+/// because they contain barriers.
+///
+/// interface Effects {
+///   /// The effect, if any, of the specified instruction.
+///   Effects::Effect effectForPhi(SILBasicBlock *)
+///
+///   /// The effect, if any, of the phis of the specified block.
+///   Effects::Effect effectForInstruction(SILInstruction *)
+/// }
+/// interface Visitor {
+///   /// Whether the indicated basic block is within the region of the graph
+///   /// that should be traversed.
+///   bool isInRegion(SILBasicBlock *block)
+///
+///   /// The roots from which all blocks in the region are backwards-reachable.
+///   ArrayRef<SILBasicBlock *> roots();
+///
+///   /// Visit each discovered begin_access instruction which defines a barrier
+///   /// scope.
+///   void visitBarrierAccessScope(BeginAccessInst *)
+/// }
+template <typename Effects, typename Visitor>
+class VisitBarrierAccessScopes {
+  using This = VisitBarrierAccessScopes<Effects, Visitor>;
+  /// Describes the effect, if any, of each instruction and phi.
+  Effects &effects;
+  /// Describes the search region and visits the found barriers.
+  Visitor &visitor;
+  /// The function in which to find barrier access scopes.
+  SILFunction *function;
+  /// The blocks which have already been visited.  Used to determine whether a
+  /// block being visited is a barrier block.
+  BasicBlockSet visited;
+
+  /// The access scopes that are live at the begin of each block after visiting
+  /// all the block's predecessors and its instructions.
+  llvm::DenseMap<SILBasicBlock *, llvm::SmallPtrSet<BeginAccessInst *, 2>>
+      liveInAccessScopes;
+  /// While visiting a block's instructions, the access scopes that are
+  /// currently live.
+  llvm::SmallPtrSet<BeginAccessInst *, 2> runningLiveAccessScopes;
+  /// The blocks in the region, in topological order.
+  SmallVector<SILBasicBlock *, 32> order;
+
+public:
+  VisitBarrierAccessScopes(SILFunction *function, Effects &effects,
+                           Visitor &visitor)
+      : effects(effects), visitor(visitor), function(function),
+        visited(function){};
+
+  /// Visit the begin_access instructions.
+  ///
+  /// Sort the backwards-reachable, in-region blocks topologically.  Then visit
+  /// them in the reverse of that order, so that when any block is visited, all
+  /// of its non-backedge* successors will already have been visited.
+  ///
+  /// While visiting, which access scopes are live within blocks is tracked,
+  /// storing the access scopes that are live at the begin of a block, and
+  /// noting that every access scope live at the begin of any successor is also
+  /// live at the end of a block.
+  ///
+  /// * These are backedges in the reversed graph, that are ignored in a reverse
+  /// depth-first search.
+  void visit() {
+    visitSortedTopologicallyBackwards(function, visitor.roots(), *this);
```

Comment:
The casual reader should just see "visit blocks in forward order" here. But it takes a lot of mental effort to figure that out. Renaming it to `visitForwardTopological` would help. potentially as a separate PR if you want

Reply:
Will fix in https://github.com/apple/swift/pull/58791.